File: services/api/app/services/maverick_client.py
"""
Maverick LLM Client - Interface to locally-hosted LLM via vLLM
Optimized for DGX/GCP enterprise deployment with vLLM OpenAI-compatible API
"""

import httpx
from typing import Optional, Dict, Any, List
import logging

logger = logging.getLogger(__name__)


class MaverickClient:
    """Client for Maverick locally-hosted LLM (vLLM server on port 8090)"""

    def __init__(self, base_url: str = "http://localhost:8090", timeout: float = 60.0):
        self.base_url = base_url
        self.timeout = timeout
        self.client = httpx.AsyncClient(timeout=timeout)

    async def close(self):
        """Close the HTTP client"""
        await self.client.aclose()

    async def __aenter__(self):
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        await self.close()

    async def completion(
        self,
        prompt: str,
        n_predict: int = 500,
        temperature: float = 0.7,
        top_p: float = 0.9,
        top_k: int = 40,
        stop: Optional[List[str]] = None,
        stream: bool = False,
    ) -> Dict[str, Any]:
        """
        Request completion from Maverick LLM (vLLM OpenAI-compatible API)

        Args:
            prompt: The input prompt
            n_predict: Max tokens to generate (default 500)
            temperature: Sampling temperature 0.0-2.0 (default 0.7)
            top_p: Nucleus sampling threshold (default 0.9)
            top_k: Top-k sampling (default 40)
            stop: Stop sequences (default ["\nUser:", "\n\n"])
            stream: Stream response (default False)

        Returns:
            {
                "content": "Generated text...",
                "tokens_predicted": 123,
                "finish_reason": "stop"
            }
        """
        if stop is None:
            stop = ["\nUser:", "\n\n"]

        # vLLM uses OpenAI-compatible /v1/completions endpoint
        payload = {
            "model": "default",  # vLLM uses the loaded model
            "prompt": prompt,
            "max_tokens": n_predict,
            "temperature": temperature,
            "top_p": top_p,
            "stop": stop,
            "stream": stream,
        }

        try:
            logger.debug(f"Sending completion request to {self.base_url}/v1/completions")
            response = await self.client.post(
                f"{self.base_url}/v1/completions", json=payload
            )
            response.raise_for_status()
            result = response.json()

            # Extract from OpenAI format
            choice = result.get("choices", [{}])[0]
            return {
                "content": choice.get("text", ""),
                "tokens_predicted": result.get("usage", {}).get("completion_tokens", 0),
                "finish_reason": choice.get("finish_reason", "unknown")
            }

        except httpx.HTTPStatusError as e:
            logger.error(
                f"HTTP error from Maverick: {e.response.status_code} - {e.response.text}"
            )
            raise Exception(f"Maverick LLM error: {e.response.status_code}")
        except httpx.RequestError as e:
            logger.error(f"Request error to Maverick: {str(e)}")
            raise Exception(f"Failed to connect to Maverick LLM: {str(e)}")
        except Exception as e:
            logger.error(f"Unexpected error calling Maverick: {str(e)}")
            raise

    async def chat(
        self,
        messages: List[Dict[str, str]],
        system_prompt: Optional[str] = None,
        temperature: float = 0.7,
        max_tokens: int = 500,
    ) -> Dict[str, Any]:
        """
        Chat-style completion with message history (vLLM OpenAI-compatible API)

        Args:
            messages: List of {role: "user"|"assistant"|"system", content: "..."}
            system_prompt: Optional system prompt (prepended to messages)
            temperature: Sampling temperature
            max_tokens: Max tokens to generate

        Returns:
            {
                "content": "Assistant response",
                "tokens_predicted": 123,
                "role": "assistant",
                "finish_reason": "stop"
            }
        """
        # Build messages list for OpenAI format
        formatted_messages = []

        if system_prompt:
            formatted_messages.append({"role": "system", "content": system_prompt})

        formatted_messages.extend(messages)

        # vLLM uses OpenAI-compatible /v1/chat/completions endpoint
        payload = {
            "model": "default",  # vLLM uses the loaded model
            "messages": formatted_messages,
            "max_tokens": max_tokens,
            "temperature": temperature,
            "top_p": 0.9,
        }

        try:
            logger.debug(f"Sending chat request to {self.base_url}/v1/chat/completions")
            response = await self.client.post(
                f"{self.base_url}/v1/chat/completions", json=payload
            )
            response.raise_for_status()
            result = response.json()

            # Extract from OpenAI format
            choice = result.get("choices", [{}])[0]
            message = choice.get("message", {})

            return {
                "content": message.get("content", "").strip(),
                "tokens_predicted": result.get("usage", {}).get("completion_tokens", 0),
                "role": "assistant",
                "finish_reason": choice.get("finish_reason", "stop"),
            }

        except httpx.HTTPStatusError as e:
            logger.error(
                f"HTTP error from Maverick: {e.response.status_code} - {e.response.text}"
            )
            raise Exception(f"Maverick LLM error: {e.response.status_code}")
        except httpx.RequestError as e:
            logger.error(f"Request error to Maverick: {str(e)}")
            raise Exception(f"Failed to connect to Maverick LLM: {str(e)}")
        except Exception as e:
            logger.error(f"Unexpected error calling Maverick: {str(e)}")
            raise

    async def health_check(self) -> bool:
        """
        Check if Maverick LLM is healthy

        Returns:
            True if healthy, False otherwise
        """
        try:
            response = await self.client.get(f"{self.base_url}/health", timeout=5.0)
            return response.status_code == 200
        except Exception as e:
            logger.warning(f"Maverick health check failed: {str(e)}")
            return False

    async def get_model_info(self) -> Dict[str, Any]:
        """
        Get model information (if endpoint exists)

        Returns:
            Model metadata dict
        """
        try:
            response = await self.client.get(f"{self.base_url}/props")
            response.raise_for_status()
            return response.json()
        except Exception as e:
            logger.warning(f"Could not fetch model info: {str(e)}")
            return {"model": "maverick", "context_length": 4096, "status": "unknown"}


# Global client instance (will be initialized on app startup)
maverick_client: Optional[MaverickClient] = None


async def get_maverick_client() -> MaverickClient:
    """Dependency for FastAPI endpoints"""
    global maverick_client
    if maverick_client is None:
        maverick_client = MaverickClient()
    return maverick_client


async def shutdown_maverick_client():
    """Shutdown client on app shutdown"""
    global maverick_client
    if maverick_client:
        await maverick_client.close()
        maverick_client = None

File: services/agents/life-navigator-training-rs/src/maverick_client.rs
//! Maverick LLM Client - Rust interface to locally-hosted Llama-4
//!
//! High-performance async client for llama.cpp server

use serde::{Deserialize, Serialize};
use reqwest::Client;
use std::time::Duration;

/// Maverick completion request
#[derive(Debug, Clone, Serialize)]
pub struct CompletionRequest {
    pub prompt: String,
    pub n_predict: u32,
    pub temperature: f64,
    pub top_p: f64,
    pub top_k: u32,
    pub stop: Vec<String>,
    pub stream: bool,
}

impl Default for CompletionRequest {
    fn default() -> Self {
        Self {
            prompt: String::new(),
            n_predict: 500,
            temperature: 0.7,
            top_p: 0.9,
            top_k: 40,
            stop: vec!["\nUser:".to_string(), "\n\n".to_string()],
            stream: false,
        }
    }
}

/// Maverick completion response
#[derive(Debug, Clone, Deserialize)]
pub struct CompletionResponse {
    pub content: String,
    pub tokens_predicted: u32,
    pub tokens_evaluated: u32,
    pub truncated: bool,
    pub stopped_eos: bool,
    pub stopped_word: bool,
    pub stopped_limit: bool,
    pub stopping_word: Option<String>,
}

/// Chat message
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChatMessage {
    pub role: String,
    pub content: String,
}

/// Chat response
#[derive(Debug, Clone)]
pub struct ChatResponse {
    pub content: String,
    pub tokens_predicted: u32,
    pub role: String,
    pub finish_reason: String,
}

/// Maverick LLM Client
#[derive(Clone)]
pub struct MaverickClient {
    client: Client,
    base_url: String,
}

impl MaverickClient {
    /// Create a new Maverick client
    pub fn new(base_url: Option<String>) -> Self {
        let base_url = base_url.unwrap_or_else(|| "http://localhost:8090".to_string());

        let client = Client::builder()
            .timeout(Duration::from_secs(60))
            .build()
            .expect("Failed to create HTTP client");

        Self { client, base_url }
    }

    /// Request completion from Maverick
    pub async fn completion(&self, request: CompletionRequest) -> Result<CompletionResponse, Box<dyn std::error::Error + Send + Sync>> {
        let url = format!("{}/completion", self.base_url);

        let response = self
            .client
            .post(&url)
            .json(&request)
            .send()
            .await?;

        if !response.status().is_success() {
            return Err(format!("Maverick error: {}", response.status()).into());
        }

        let result: CompletionResponse = response.json().await?;
        Ok(result)
    }

    /// Chat with Maverick using message history
    pub async fn chat(
        &self,
        messages: Vec<ChatMessage>,
        system_prompt: Option<String>,
        temperature: f64,
        max_tokens: u32,
    ) -> Result<ChatResponse, Box<dyn std::error::Error + Send + Sync>> {
        // Build prompt from messages
        let mut prompt_parts = Vec::new();

        if let Some(system) = system_prompt {
            prompt_parts.push(format!("System: {}\n", system));
        }

        for msg in messages {
            let role = capitalize_first(&msg.role);
            prompt_parts.push(format!("{}: {}", role, msg.content));
        }

        prompt_parts.push("Assistant:".to_string());
        let full_prompt = prompt_parts.join("\n\n");

        // Call completion
        let request = CompletionRequest {
            prompt: full_prompt,
            n_predict: max_tokens,
            temperature,
            top_p: 0.9,
            top_k: 40,
            stop: vec!["\nUser:".to_string(), "\nSystem:".to_string(), "\n\n".to_string()],
            stream: false,
        };

        let response = self.completion(request).await?;

        // Format as chat response
        let finish_reason = if response.stopped_word || response.stopped_eos {
            "stop"
        } else {
            "length"
        };

        Ok(ChatResponse {
            content: response.content.trim().to_string(),
            tokens_predicted: response.tokens_predicted,
            role: "assistant".to_string(),
            finish_reason: finish_reason.to_string(),
        })
    }

    /// Health check
    pub async fn health_check(&self) -> bool {
        let url = format!("{}/health", self.base_url);

        match self.client.get(&url).timeout(Duration::from_secs(5)).send().await {
            Ok(response) => response.status().is_success(),
            Err(_) => false,
        }
    }

    /// Get model info
    pub async fn model_info(&self) -> Result<serde_json::Value, Box<dyn std::error::Error + Send + Sync>> {
        let url = format!("{}/props", self.base_url);

        let response = self.client.get(&url).send().await?;

        if !response.status().is_success() {
            return Ok(serde_json::json!({
                "model": "maverick",
                "context_length": 4096,
                "status": "unknown"
            }));
        }

        let info = response.json().await?;
        Ok(info)
    }
}

fn capitalize_first(s: &str) -> String {
    let mut chars = s.chars();
    match chars.next() {
        None => String::new(),
        Some(first) => first.to_uppercase().collect::<String>() + chars.as_str(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_completion_request_default() {
        let request = CompletionRequest::default();
        assert_eq!(request.n_predict, 500);
        assert_eq!(request.temperature, 0.7);
        assert!(!request.stream);
    }

    #[test]
    fn test_chat_message() {
        let msg = ChatMessage {
            role: "user".to_string(),
            content: "Hello".to_string(),
        };
        assert_eq!(msg.role, "user");
    }

    #[test]
    fn test_capitalize_first() {
        assert_eq!(capitalize_first("user"), "User");
        assert_eq!(capitalize_first("assistant"), "Assistant");
        assert_eq!(capitalize_first(""), "");
    }
}
